File: DesignPatternsPractices/DesignPatternsPractices/DesignPatterns/Decorator_CPP/Decorator.cpp
//
//  Decorator.cpp
//  DesignPatternsPractices
//
//  Created by p2pmsg on 14-8-14.
//  Copyright (c) 2014年 com.skyprayer. All rights reserved.
//

#include "Decorator.h"
#include <stdio.h>


Decorator::Decorator()
{
    
}


Decorator::~Decorator()
{
    
}


void Decorator::operation()
{
    mComponent->operation();
}


void Decorator::setComponent(Component* pComponent)
{
    this->mComponent = pComponent;
}


ConcreteDecorator::ConcreteDecorator()
{
    
}


ConcreteDecorator::~ConcreteDecorator()
{
    
}



void ConcreteDecorator::addBehavior()
{
    printf("ConcreteDecorator's addBehavior!\n");
}

File: Design_Patterns/DesignPatternsPractices/DesignPatternsPractices/DesignPatterns/Structural/Decorator/CodeTemplate/Decorator.m
//
//  Decorator.m
//  DesignPatternsPractices
//
//  Created by su xinde on 14-9-9.
//  Copyright (c) 2014年 Su XinDe. All rights reserved.
//

#import "Decorator.h"

@implementation Decorator

- (void)setComponent:(Decorator_Component *)component
{
    if (mComponent) {
        [mComponent release];
        mComponent = nil;
    }
    mComponent = [component retain];
    
}

- (void)operation
{
    if (mComponent != nil) {
        [mComponent operation];
    }
}



@end
